### PR TITLE
 Refactor transaction storage pallet to use fungible traits

### PR DESCRIPTION
**Pull request by acatangiu**
Partial https://github.com/paritytech/polkadot-sdk/issues/226

`frame/transaction-storage`: replace `Currency` with `fungible::*` traits

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3885728

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3886072

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/3896342

File: substrate/frame/transaction-storage/src/lib.rs [hunk 1]
```diff
@@ -98,10 +109,13 @@ pub mod pallet {
 			+ Dispatchable<RuntimeOrigin = Self::RuntimeOrigin>
 			+ GetDispatchInfo
 			+ From<frame_system::Call<Self>>;
-		/// The currency trait.
-		type Currency: ReservableCurrency<Self::AccountId>;
+		/// The currency type for this pallet.
+		type Currency: FnMutateHold<Self::AccountId, Reason = Self::RuntimeHoldReason>
```
Comment (liamaharon) [inline substrate/frame/transaction-storage/src/lib.rs:113]: see discussion https://github.com/paritytech/polkadot-sdk/issues/226#issuecomment-1691692395

```suggestion
		/// The fungible type for this pallet.
		type Fungible: FnMutateHold<Self::AccountId, Reason = Self::RuntimeHoldReason>
```

File: substrate/frame/transaction-storage/src/lib.rs [hunk 2]
```diff
@@ -98,10 +109,13 @@ pub mod pallet {
...
+			+ FnBalanced<Self::AccountId>;
+		/// The overarching runtime hold reason.
```
Comment (liamaharon) [inline substrate/frame/transaction-storage/src/lib.rs:115]:
```suggestion
			+ FnBalanced<Self::AccountId>;

		/// The overarching runtime hold reason.
```

File: substrate/frame/transaction-storage/src/lib.rs [hunk 3]
```diff
@@ -432,8 +444,8 @@ pub mod pallet {
 			let byte_fee = ByteFee::<T>::get().ok_or(Error::<T>::NotConfigured)?;
 			let entry_fee = EntryFee::<T>::get().ok_or(Error::<T>::NotConfigured)?;
 			let fee = byte_fee.saturating_mul(size.into()).saturating_add(entry_fee);
-			ensure!(T::Currency::can_slash(&sender, fee), Error::<T>::InsufficientFunds);
-			let (credit, _) = T::Currency::slash(&sender, fee);
+			T::Currency::hold(&HoldReason::StorageFeeHold.into(), &sender, fee)?;
+			let (credit, _) = T::Currency::slash(&HoldReason::StorageFeeHold.into(), &sender, fee);
```
Comment (liamaharon) [inline substrate/frame/transaction-storage/src/lib.rs:448]: Seems logically equivalent to just reducing the account balance by `fee` if the balance is gt `fee`? Or am I missing something?

File: substrate/frame/transaction-storage/src/benchmarking.rs [hunk 4]
```diff
@@ -129,7 +126,8 @@ benchmarks! {
 	store {
 		let l in 1 .. T::MaxTransactionSize::get();
 		let caller: T::AccountId = whitelisted_caller();
-		T::Currency::make_free_balance_be(&caller, BalanceOf::<T>::max_value());
+		let initial_balance = BalanceOf::<T>::max_value().checked_div(&2u32.into()).unwrap();
```
Comment (juangirini) [inline substrate/frame/transaction-storage/src/benchmarking.rs:129]: We don't need a `checked_div` when we know we aren't dividing by 0

File: substrate/frame/transaction-storage/src/benchmarking.rs [hunk 5]
```diff
@@ -138,7 +136,8 @@ benchmarks! {
 
 	renew {
 		let caller: T::AccountId = whitelisted_caller();
-		T::Currency::make_free_balance_be(&caller, BalanceOf::<T>::max_value());
+		let initial_balance = BalanceOf::<T>::max_value().checked_div(&2u32.into()).unwrap();
```
Comment (juangirini) [inline substrate/frame/transaction-storage/src/benchmarking.rs:139]: same

File: substrate/frame/transaction-storage/src/benchmarking.rs [hunk 6]
```diff
@@ -152,7 +151,8 @@ benchmarks! {
 	check_proof_max {
 		run_to_block::<T>(1u32.into());
 		let caller: T::AccountId = whitelisted_caller();
-		T::Currency::make_free_balance_be(&caller, BalanceOf::<T>::max_value());
+		let initial_balance = BalanceOf::<T>::max_value().checked_div(&2u32.into()).unwrap();
```
Comment (juangirini) [inline substrate/frame/transaction-storage/src/benchmarking.rs:154]: same here

File: substrate/frame/transaction-storage/src/lib.rs [hunk 7]
```diff
@@ -98,10 +110,14 @@ pub mod pallet {
 			+ Dispatchable<RuntimeOrigin = Self::RuntimeOrigin>
 			+ GetDispatchInfo
 			+ From<frame_system::Call<Self>>;
-		/// The currency trait.
-		type Currency: ReservableCurrency<Self::AccountId>;
+		/// The fungible type for this pallet.
+		type Fungible: FnMutate<Self::AccountId>
+			+ FnMutateHold<Self::AccountId, Reason = Self::RuntimeHoldReason>
+			+ FnBalanced<Self::AccountId>;
```
Comment (juangirini) [inline substrate/frame/transaction-storage/src/lib.rs:116]: I am in favour of sticking to `Currency` as the naming. I know there is no consensus on this though.
The rational behind is that this type still represents a currency even though it does not directly implement a trait called currency

File: substrate/frame/transaction-storage/src/mock.rs [hunk 8]
```diff
@@ -68,26 +71,20 @@ impl frame_system::Config for Test {
 	type MaxConsumers = ConstU32<16>;
 }
 
+#[derive_impl(pallet_balances::config_preludes::TestDefaultConfig as pallet_balances::DefaultConfig)]
```
Comment (juangirini) [inline substrate/frame/transaction-storage/src/mock.rs:74]: I would usually prefer unrelated changes to be in a separate PR

**Review by juangirini — APPROVED** [review]
LGTM just a few optional nits

File: substrate/frame/transaction-storage/src/lib.rs (hunk 3 above)
Comment (juangirini) [reply substrate/frame/transaction-storage/src/lib.rs]: You could very well use `decrease_balance` which returns the reduced balance. I don't see why that wouldn't work.
Though, I don't know why it was not used in the original code

File: substrate/frame/transaction-storage/src/benchmarking.rs (hunk 4 above)
Comment (acatangiu) [reply substrate/frame/transaction-storage/src/benchmarking.rs:129]: this is just benchmark setup, no real point in optimizing.. I prefer to leave it like this (defensive even if useless)

File: substrate/frame/transaction-storage/src/lib.rs (hunk 7 above)
Comment (acatangiu) [reply substrate/frame/transaction-storage/src/lib.rs:116]: done

File: substrate/frame/transaction-storage/src/mock.rs (hunk 8 above)
Comment (acatangiu) [reply substrate/frame/transaction-storage/src/mock.rs:74]: neah, it's small change that can go here easily and avoid another PR

**Comment by acatangiu** [comment]
@liamaharon your comments are addressed, please take another look 🙏

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable 1/3
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/4124271

File: substrate/frame/transaction-storage/src/lib.rs [hunk 9]
```diff
@@ -432,8 +446,8 @@ pub mod pallet {
 			let byte_fee = ByteFee::<T>::get().ok_or(Error::<T>::NotConfigured)?;
 			let entry_fee = EntryFee::<T>::get().ok_or(Error::<T>::NotConfigured)?;
 			let fee = byte_fee.saturating_mul(size.into()).saturating_add(entry_fee);
-			ensure!(T::Currency::can_slash(&sender, fee), Error::<T>::InsufficientFunds);
-			let (credit, _) = T::Currency::slash(&sender, fee);
+			T::Currency::hold(&HoldReason::StorageFeeHold.into(), &sender, fee)?;
+			let (credit, _) = T::Currency::slash(&HoldReason::StorageFeeHold.into(), &sender, fee);
```
Comment (georgepisaltu) [inline substrate/frame/transaction-storage/src/lib.rs:450]: nit: `T::Currency::hold` must ensure there is no slash remainder, but just for tests
```suggestion
			let (credit, remainder) = T::Currency::slash(&HoldReason::StorageFeeHold.into(), &sender, fee);
			debug_assert!(remainder.is_zero());
```

**Review by georgepisaltu — APPROVED** [review]

**Comment by gupnik** [comment]
@acatangiu Any reason why `Consideration` was not used here? Seems perfectly suitable to this use-case.
